Sources/PostgresNIO/New/Connection State Machine/AuthenticationStateMachine.swift

```diff
@@ -62,33 +66,86 @@ struct AuthenticationStateMachine {
                 return self.setAndFireError(.unsupportedAuthMechanism(.gss))
             case .sspi:
                 return self.setAndFireError(.unsupportedAuthMechanism(.sspi))
-            case .sasl:
-                return self.setAndFireError(.unsupportedAuthMechanism(.sasl))
+            case .sasl(let mechanisms):
+                guard mechanisms.contains("SCRAM-SHA-256") else {
+                    return self.setAndFireError(.unsupportedAuthMechanism(.sasl(mechanisms: mechanisms)))
+                }
+                
+                guard let password = self.authContext.password else {
+                    return self.setAndFireError(.authMechanismRequiresPassword)
+                }
+                
+                let saslManager = SASLAuthenticationManager(asClientSpeaking:
+                    SASLMechanism.SCRAM.SHA256(username: self.authContext.username, password: { password }))
+                
+                do {
+                    var bytes: [UInt8]?
+                    let done = try saslManager.handle(message: nil, sender: { bytes = $0 })
+                    
+                    guard let output = bytes, done == false else {
```

Comment:
Can we add a TODO comment here of the form ```swift
// TODO: Gwynne reminds herself to refactor `SASLAuthenticationManager` to be async instead of very badly done synchronous.
```